### PR TITLE
Filter and report generation

### PR DESCRIPTION
**Pull request by frapzo**
Added features:

* multi-threaded crawling for better performance using multiprocessing.Pool
* modified identity of crawler to bypass most filters
* wikipedia search
* filter for collected web pages
* report generation by cross referencing filtered web pages

### CODE REVIEW

**Comment by frapzo** [comment]
Example of resulting report when searching for microsoft:

[report.md](https://github.com/SondreUM/gen-ai/files/15191235/report.md)

**Review by stiansolli — APPROVED** [review]
lgtm, nicely done!

**Comment by frapzo** [comment]
some bugs were discovered with excessively large lines when running the filter. so will close down the pull request until it's fixed

**Comment by frapzo** [comment]
filter will also save anything rejected by Azure content filter to a review directory for manual review

File: src/processing.py [hunk 1]
```diff
@@ -26,19 +30,52 @@ def run_crawler(urls: list[str], max_depth: int):
...
 
 
+def search_wikipedia(query: str) -> None:
+    """Search wikipedia for the query"""
+    crawler_dir: Path = Path(config.DATA_PATH).joinpath("crawler_data")
+    modified_query = query.replace(" ", "%20")
+    modified_query = modified_query.replace("&", "%26")
+
+    # search for appropriate wikipedia page (limit defines how many results to return)
+    response = get(f"https://api.wikimedia.org/core/v1/wikipedia/en/search/title?q={modified_query}&limit=1")
```
Comment (stiansolli) [inline src/processing.py:40]: Should probably utilize the same gpt rankings scheme as the google search on wikipedia, having only the top result may lead to wrong results. As in the case of searching for UiT, the top result from this is UiTM United Football Club...

File: src/filter.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                    # handle excessively large lines
+                    to_read = 3000 - tokens
+                    tokens = 0
+                    data += new_line[:to_read]
+                    new_line = new_line[to_read:]
+                    gpt_response = use_gpt(data, organization, file)
+                    if len(gpt_response) > 0:
+                        result += f"\n{gpt_response}"
+                    while len(new_line) > 0:
+                        to_read = len(new_line) if len(new_line) // 4 < 3000 else 3000*4
```
Comment (stiansolli) [inline src/filter.py:75]: Why else 3000*4 and not 3000 as before?

File: src/filter.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+                if len(new_line) == 0:
+                    # end of file
+                    gpt_response = use_gpt(data, organization, file)
+                    if len(gpt_response) > 0:
+                        result += f"\n{gpt_response}"
+                    break
+                elif tokens > 2500:
+                    # approaching token limit
+                    data += new_line
+                    gpt_response = use_gpt(data, organization, file)
```
Comment (stiansolli) [inline src/filter.py:90]: I think getting GPT to summarize with slicing this way may yield some weird results, however I know that this is an edgecase based on the token limit of gpt.

**Review by stiansolli — CHANGES_REQUESTED** [review]
See comments.

File: src/filter.py (hunk 2 above)
Comment (frapzo) [reply src/filter.py:75]: to_read is specifically how many characters to read. Not how many tokens to read. as such 3000*4 means that we read 3000 tokens

File: src/filter.py (hunk 3 above)
Comment (frapzo) [reply src/filter.py:90]: This is a compromise that will essentially always exist as long as the model used is one with a token limit. It sucks, but that is just reality. with gpt we are also limited by the token limit of the reponses, so even if the input token limit was eliminated one would want to slice it up due to response limit.

**Pull request merged**